pkg/problemmetrics/problem_metrics.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package problemmetrics
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/node-problem-detector/pkg/util/metrics"
+)
+
+// GlobalProblemMetricsManager is a singleton of ProblemMetricsManager,
+// which should be used to manage all problem-converted metrics across all
+// problem daemons.
+var GlobalProblemMetricsManager *ProblemMetricsManager
+
+func init() {
+	GlobalProblemMetricsManager = NewProblemMetricsManagerOrDie()
+}
+
+// ProblemMetricsManager manages problem-converted metrics.
+// ProblemMetricsManager is thread-safe.
+type ProblemMetricsManager struct {
+	problemCounter         metrics.Int64MetricInterface
+	problemGauge           metrics.Int64MetricInterface
+	conditionToReason      map[string]string
+	conditionToReasonMutex sync.Mutex
+}
+
+func NewProblemMetricsManagerOrDie() *ProblemMetricsManager {
+	pmm := ProblemMetricsManager{}
+
+	var err error
+	pmm.problemCounter, err = metrics.NewInt64Metric(
+		"problem_counter",
+		"Number of times a specific type of problem have occurred.",
+		"1",
+		metrics.Sum,
```

Comment:
Should this be sum or count?

Reply:
I slightly prefer Sum for following reasons:
1. Sum's operations is a super set of Count's operations.
2. They will both show up as `# TYPE counter` in Prometheus metrics, so they won't confuse the user.
3. If we use Sum, we retain the flexibility to implement below behaviors (I'm not saying we want to implement these behaviors. I just think that **if** we want them for some reason, we should be able to do them):

    - Preset a data point for the metrics during initialization, with value as 0.
All metrics are created empty, with no data point. `Sum(0)` will introduce a 0-valued data point, while `Count(0)` will introduce a 1-valued data point. There is no way to introduce a 0-valued Sum data point.
    - Make the data point stateful across NPD restarts/crashes.
Today, when NPD crashes, all metrics are lost, and will start from 0 after NPD starts up.
Say **if** we want to allow NPD to store the counters at disk, and restore them after crash. And if the counter has a value of 1000, it would require calling `Count(x)` 1000 times, while require calling `Sum(1000)` one time.
Using Count make implementing this much harder.

Reply:
Thanks for the explanation!